File: app/src/main/java/com/sortedqueue/programmercreek/fragments/PresentationsListFragment.java
package com.sortedqueue.programmercreek.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.firebase.database.DatabaseError;
import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.activity.CreatePresentationActivity;
import com.sortedqueue.programmercreek.activity.ViewPresentationActivity;
import com.sortedqueue.programmercreek.adapter.CustomProgramRecyclerViewAdapter;
import com.sortedqueue.programmercreek.adapter.PresentationsListRecyclerAdapter;
import com.sortedqueue.programmercreek.constants.ProgrammingBuddyConstants;
import com.sortedqueue.programmercreek.database.PresentationModel;
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Alok Omkar on 2017-04-04.
 */
public class PresentationsListFragment extends Fragment implements View.OnClickListener, FirebaseDatabaseHandler.GetAllPresentationsListener, CustomProgramRecyclerViewAdapter.AdapterClickListner {

    private static PresentationsListFragment instance;
    @BindView(R.id.presentationsRecyclerView)
    RecyclerView presentationsRecyclerView;
    private PresentationsListRecyclerAdapter adapter;


    public static PresentationsListFragment getInstance() {
        if (instance == null) {
            instance = new PresentationsListFragment();
        }
        return instance;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_presentations, container, false);
        ButterKnife.bind(this, view);
        new FirebaseDatabaseHandler(getContext()).getAllPresentations(this);
        return view;
    }

    private void setupRecyclerView(ArrayList<PresentationModel> presentationModelArrayList) {
        presentationsRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2));
        adapter = new PresentationsListRecyclerAdapter(getContext(), presentationModelArrayList, this);
        presentationsRecyclerView.setAdapter(adapter);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

    }

    @Override
    public void onClick(View v) {
        Intent intent = new Intent(getContext(), CreatePresentationActivity.class);
        startActivity(intent);
    }

    @Override
    public void onSuccess(ArrayList<PresentationModel> presentationModelArrayList) {
        setupRecyclerView(presentationModelArrayList);
    }

    @Override
    public void onError(DatabaseError databaseError) {

    }

    @Override
    public void onItemClick(int position) {
        PresentationModel presentationModel = adapter.getItemAtPosition(position);
        Intent intent = new Intent(getContext(), ViewPresentationActivity.class);
        intent.putExtra(ProgrammingBuddyConstants.INSTANCE.getKEY_PROG_ID(), presentationModel);
        startActivity(intent);
    }
}

File: app/src/main/java/com/sortedqueue/programmercreek/fragments/PresentationsListFragment.kt
package com.sortedqueue.programmercreek.fragments

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.firebase.database.DatabaseError
import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.activity.CreatePresentationActivity
import com.sortedqueue.programmercreek.activity.ViewPresentationActivity
import com.sortedqueue.programmercreek.adapter.CustomProgramRecyclerViewAdapter
import com.sortedqueue.programmercreek.adapter.PresentationsListRecyclerAdapter
import com.sortedqueue.programmercreek.constants.ProgrammingBuddyConstants
import com.sortedqueue.programmercreek.database.PresentationModel
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler
import kotlinx.android.synthetic.main.fragment_presentations.*

import java.util.ArrayList




/**
 * Created by Alok Omkar on 2017-04-04.
 */
class PresentationsListFragment : Fragment(), View.OnClickListener, FirebaseDatabaseHandler.GetAllPresentationsListener, CustomProgramRecyclerViewAdapter.AdapterClickListner {

    private var adapter: PresentationsListRecyclerAdapter? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater!!.inflate(R.layout.fragment_presentations, container, false)


        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        FirebaseDatabaseHandler(context).getAllPresentations(this)
    }

    private fun setupRecyclerView(presentationModelArrayList: ArrayList<PresentationModel>) {
        presentationsRecyclerView!!.layoutManager = GridLayoutManager(context, 2)
        adapter = PresentationsListRecyclerAdapter(context, presentationModelArrayList, this)
        presentationsRecyclerView!!.adapter = adapter
    }

    override fun onDestroyView() {
        super.onDestroyView()

    }

    override fun onClick(v: View) {
        val intent = Intent(context, CreatePresentationActivity::class.java)
        startActivity(intent)
    }

    override fun onSuccess(presentationModelArrayList: ArrayList<PresentationModel>) {
        setupRecyclerView(presentationModelArrayList)
    }

    override fun onError(databaseError: DatabaseError?) {

    }

    override fun onItemClick(position: Int) {
        val presentationModel = adapter!!.getItemAtPosition(position)
        val intent = Intent(context, ViewPresentationActivity::class.java)
        intent.putExtra(ProgrammingBuddyConstants.KEY_PROG_ID, presentationModel)
        startActivity(intent)
    }

    companion object {

        @SuppressLint("StaticFieldLeak")
        private var indexFragment: PresentationsListFragment? = null
        val instance: PresentationsListFragment
            get() {
                if (indexFragment == null) {
                    indexFragment = PresentationsListFragment()
                }
                return indexFragment!!
            }
    }
}
